### ISSUE TITLE
Issue where NeuralNet model is not resetting weights for each k-fold in Scikit-Learn's cross_validate.

### ISSUE THREAD

**Issue by tjens150** [skorch-dev/skorch#1073]
Hi,

I was attempting to use Skorch's "NeuralNetBinaryClassifier" to use Scikit-Learn's cross_validate on a simple MLP for classification. I noticed that when I set verbose=1, the train_loss was significantly lower at start for each k-fold, and the metrics were too good to be true.

I believe the model's weights are NOT being reset between training on each fold despite turning warm_start=False and train_split=None.

```
import torch
import torch.nn as nn
from torch.optim.adamw import AdamW
from skorch import NeuralNetBinaryClassifier
from sklearn.model_selection import cross_validate

class MLP(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = torch.sigmoid(self.fc2(x))
        return x.to(device)

input_size=771
hidden_size=100
output_size=1
MLP_args=(input_size,hidden_size,output_size)
MLP_kwargs={'module' : MLP(*MLP_args),'criterion' : nn.BCELoss, 'optimizer' : AdamW,'lr' : 0.001, 'max_epochs' : 40, 'device' : device, 'verbose' : 1}
model=NeuralNetBinaryClassifier(train_split=None,warm_start=False,**MLP_kwargs))
cross_validate(model,x_train,y_train)
```
```
  epoch    train_loss     dur
-------  ------------  ------
      1        0.6594  0.1411
      2        0.6164  0.0908
      3        0.5827  0.1130
      4        0.5560  0.1225
      5        0.5413  0.0841
      6        0.5257  0.1130
      7        0.5154  0.0900
      8        0.5070  0.1061
      9        0.4977  0.1093
     10        0.4906  0.1383
     11        0.4840  0.1164
     12        0.4769  0.1210
     13        0.4710  0.1077
     14        0.4652  0.1072
     15        0.4596  0.0657
     16        0.4536  0.0599
     17        0.4487  0.0906
     18        0.4434  0.0552
     19        0.4383  0.0542
     20        0.4335  0.0628
     21        0.4294  0.0729
     22        0.4244  0.1212
     23        0.4196  0.1273
     24        0.4158  0.1146
     25        0.4116  0.1241
     26        0.4077  0.1156
     27        0.4034  0.0949
     28        0.3999  0.0517
     29        0.3958  0.0595
     30        0.3921  0.1199
     31        0.3885  0.1137
     32        0.3847  0.1138
     33        0.3813  0.1071
     34        0.3777  0.0595
     35        0.3742  0.0948
     36        0.3709  0.0827
     37        0.3677  0.1159
     38        0.3647  0.1126
     39        0.3615  0.1229
     40        0.3580  0.1105
  epoch    train_loss     dur
-------  ------------  ------
      1        0.3549  0.0673
      2        0.3420  0.1037
      3        0.3734  0.0983
      4        0.3455  0.1046
      5        0.3330  0.0646
      6        0.3401  0.1072
      7        0.3257  0.1069
      8        0.3285  0.1151
      9        0.3253  0.0634
     10        0.3192  0.0568
     11        0.3199  0.1064
...
```

Is this the intended behavior?

I was able to fix it by adding a callback that resets the weights manually, but I don't think this is expected behavior.

```
from skorch.callbacks import Callback

class ResetParametersCallback(Callback):
    """Custom callback to reset model parameters at the beginning of training."""
    def on_train_begin(self, net, **kwargs):
        # Reset parameters for all layers that implement 'reset_parameters'
        net.module_.apply(
            lambda module: module.reset_parameters() if hasattr(module, 'reset_parameters') else None
        )
MLP_kwargs={'module' : MLP(*MLP_args),'criterion' : nn.BCELoss, 'optimizer' : AdamW,'lr' : 0.001, 'max_epochs' : 40, 'device' : device, 'verbose' : 1,'callbacks' : [ResetParametersCallback()]}
NeuralNetBinaryClassifier(train_split=None,warm_start=False,**MLP_kwargs))
```

**Comment by BenjaminBossan** [comment]
Thanks for the detailed report. Your intuition is correct, the weights should be reset for each cross validation run. Since the provided code is not fully complete, I made a few additions to make it run but otherwise stuck very close to your code. What I found is that the model is correctly reset on each split. As you can see below, the train loss always starts at ~0.70, which is expected for a binary classification task with equal number of 0s and 1s.

```python
import numpy as np
import torch
import torch.nn as nn
from torch.optim.adamw import AdamW
from skorch import NeuralNetBinaryClassifier
from sklearn.model_selection import cross_validate
from sklearn.datasets import make_classification

np.random.seed(0)
torch.manual_seed(0)

device = "cuda:0"

class MLP(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = torch.sigmoid(self.fc2(x))
        return x.to(device)

input_size = 20
hidden_size = 100
output_size = 1
MLP_args=(input_size,hidden_size,output_size)
MLP_kwargs={
    'module' : MLP(*MLP_args),
    'criterion' : nn.BCELoss,
    'optimizer' : AdamW,
    'lr' : 0.001,
    'max_epochs' : 40,
    'device' : device,
    'verbose' : 1,
}

X_train, y_train = make_classification()
X_train = X_train.astype(np.float32)
y_train = y_train.astype(np.float32)
model = NeuralNetBinaryClassifier(train_split=None, warm_start=False, **MLP_kwargs)
cross_validate(model, X_train, y_train)
```

Output:

```
  epoch    train_loss     dur
-------  ------------  ------
      1        0.7037  0.1209
      2        0.6939  0.0013
      3        0.6843  0.0009
      4        0.6749  0.0010
      5        0.6658  0.0009
...
     35        0.4440  0.0008
     36        0.4376  0.0008
     37        0.4312  0.0008
     38        0.4249  0.0008
     39        0.4187  0.0008
     40        0.4124  0.0008

  epoch    train_loss     dur
-------  ------------  ------
      1        0.6966  0.0008
...
     40        0.4241  0.0008

  epoch    train_loss     dur
-------  ------------  ------
      1        0.7039  0.0008
...
     40        0.4284  0.0008

  epoch    train_loss     dur
-------  ------------  ------
      1        0.7087  0.0008
...
     40        0.4045  0.0008

  epoch    train_loss     dur
-------  ------------  ------
      1        0.7042  0.0008
...
     40        0.4182  0.0008

{'fit_time': array([0.72137332, 0.04253745, 0.04591966, 0.0435977 , 0.0412817 ]),
 'score_time': array([0.00096679, 0.00072622, 0.00082898, 0.00069809, 0.00072885]),
 'test_score': array([0.75, 0.9 , 0.8 , 0.6 , 0.75])}
```

It's unclear to me why you see different results, it could just be bad luck with the splits or you're using some older version of one of the libraries. Could you try updating? If that doesn't help, could you share the versions you use (skorch, sklearn, torch)?
